### PR TITLE
Refine event system

### PR DESCRIPTION
**Pull request by WenhaiLin**

### CODE REVIEW

**Comment by WenhaiLin** [comment]
@dumganhar Please review this PR.

File: cocos/base/CCEvent.cpp [hunk 1]
```diff
@@ -31,6 +32,27 @@ Event::Event(Type type)
...
+            _listenerID = EventListener::LISTENER_ID_MOUSE;
+            break;
+        case Event::Type::FOCUS:
+            _listenerID = EventListener::LISTENER_ID_FOCUS;
+            break;
+        case Event::Type::GAME_CONTROLLER:
+            _listenerID = EventListener::LISTENER_ID_GAME_CONTROLLER;
+            break;
+        default:
+            _listenerID = EventListener::LISTENER_ID_UNKNOWN;
```
Comment (dumganhar) [inline cocos/base/CCEvent.cpp]: It's strange to putting a `_listenerID` into `Event` class.

File: cocos/base/CCEventListener.cpp [hunk 2]
```diff
@@ -52,5 +85,15 @@ bool EventListener::checkAvailable()
     return (_onEvent != nullptr);
 }
 
+EventListener::ListenerID EventListener::getCustomListenerID(const std::string& eventName)
+{
+    static std::hash<std::string> h;
+    if (eventName.empty()) {
+        return EventListener::LISTENER_ID_CUSTOM;
+    } else {
+        return h(eventName) + EventListener::LISTENER_ID_CUSTOM;
```
Comment (dumganhar) [inline cocos/base/CCEventListener.cpp]: Is it possible that `EventListener::LISTENER_ID_CUSTOM` may cause overflow of `size_t`?

File: cocos/base/CCEvent.cpp (hunk 1 above)
Comment (WenhaiLin) [inline cocos/base/CCEvent.cpp]: What do you suggest?

File: cocos/base/CCDirector.cpp [hunk 3]
```diff
@@ -306,7 +318,10 @@ void Director::drawScene()
     }
     _renderer->render();
 
-    _eventDispatcher->dispatchEvent(_eventAfterDraw);
+    auto listenerCount = _eventDispatcher->getListenerCount(_eventAfterDraw->getEventListenerID());
+    if (listenerCount > 0) {
+        _eventDispatcher->dispatchEvent(_eventAfterDraw);
+    }
```
Comment (dumganhar) [inline cocos/base/CCDirector.cpp]: Please put these logic to the internal of EventDispatcher::dispatchEvent.

**Pull request merged**